compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3403,146 +3403,441 @@ TR::Register *OMR::X86::TreeEvaluator::encodeUTF16Evaluator(TR::Node *node, TR::
+   //    Note here we didn't pack a 8 bytes register because there will be only
+   //       one move for that 8 byte register, so the packing cost is too big for
+   //       just a single move.
+   // Else we pack an XMM first, and we handle the reminder using a "shifting window" method
+   //    e.g. array size is 17 byte, element size is 1 byte, here is the psudocode:
+   //    1. Pack an XMM register, xmmA, using PSHUFB
+   //    2. Store xmmA to memory at array start address
+   //    3. Store xmmA to memory at array start address + 1
+
+   if (totalSize < 16)
+      {
+      for (int i=1; i<5; i++)
+         {
+         if (packs[i] > elementSize && moves[i] <= notWorthPacking)
+            {
+            moves[i+1] += moves[i]*2;
+            moves[i] = 0;
+            }
+         else
+            {
+            break;
+            }
+         }
+      // Start packing
+      TR::Register* currentReg = valueReg;
+      for (int i=1; i<5; i++)
+         {
+         if (moves[i] > 0 && packs[i] > elementSize)
+            {
+            int currentSize = elementSize;
+            TR::Register* tempReg = cg->allocateRegister();
+            while(currentSize < packs[i])
+               {
+               PackUseShift(node, tempReg, currentReg, currentSize, cg);
+               currentSize *= 2;
+               }
+            cg->stopUsingRegister(tempReg);
+            break;
+            }
+         }
+
+      // Start moving
+      int index = 0;
+      for (int i=1;i<5;i++)
+         {
+         for (int j=0; j<moves[i]; j++)
+            {
+            GenerateMoveInstruction(node, currentReg, packs[i], addressReg, index, cg);
+            index += packs[i];
+            }
+         }
+      }
+   else
+      {
+      // First check if we are able to do it by just doing equal or less than notWorthPacking movs of the valueReg
+      if (size < notWorthPacking)
+         {
+         for (int i=0; i<size; i++)
+            {
+            GenerateMoveInstruction(node, valueReg, elementSize, addressReg, i*elementSize, cg);
+            }
+         }
+      else
+         {
+         TR::Register* XMM = cg->allocateRegister(TR_FPR);
+         PackXMM(node, XMM, valueReg, elementSize, cg);
+
+         for (int i=0; i<moves[0]; i++)
+            {
+            GenerateMoveInstruction(node, XMM, 16, addressReg, i*16, cg);
+            }
+         const int reminder = totalSize - moves[0]*16;
+         if (reminder == elementSize)
+            {
+            GenerateMoveInstruction(node, valueReg, elementSize, addressReg, moves[0]*16, cg);
+            }
+         else if (reminder != 0)
+            {
+            GenerateMoveInstruction(node, XMM, 16, addressReg, totalSize-16, cg);
+            }
+         cg->stopUsingRegister(XMM);
+         }
+      }
+   }
+
+inline void ArraySet64BitPrimitiveOn32BitISA(TR::Node* node, TR::Register* addressReg, TR::Register* valueReg, TR::Register* sizeReg, TR::CodeGenerator* cg)
```

Comment:
* Please elaborate on why this function is marked `inline` rather than `static`?

* I'd drop the `ISA` on the function name.  If you really want to say `32BitISA` then say `IA32` instead for consistency with the rest of the codegen.